src/webgpu/api/operation/rendering/basic.spec.ts

```diff
@@ -129,4 +137,191 @@ g.test('large_draw')
+          }),
+          entryPoint: 'main',
+          targets: [{ format: 'rgba8unorm' }],
+        },
+        primitive: { topology: 'point-list' },
+      });
+
+      const encoder = t.device.createCommandEncoder();
+      const pass = encoder.beginRenderPass({
+        colorAttachments: [
+          {
+            view: colorAttachmentView,
+            storeOp: 'store',
+            loadValue: { r: 1.0, g: 0.0, b: 0.0, a: 1.0 },
+          },
+        ],
+      });
+
+      pass.setPipeline(pipeline);
+      if (indexed) {
+        pass.setIndexBuffer(indexBuffer, 'uint32');
+      }
+
+      if (indirect) {
+        writeIndirectParams(numVertices, numInstances);
+        if (indexed) {
+          pass.drawIndexedIndirect(indirectBuffer, 0);
+        } else {
+          pass.drawIndirect(indirectBuffer, 0);
+        }
+      } else {
+        if (indexed) {
+          pass.drawIndexed(numVertices, numInstances);
+        } else {
+          pass.draw(numVertices, numInstances);
+        }
+      }
+      pass.endPass();
+      encoder.copyTextureToBuffer(
+        { texture: colorAttachment, mipLevel: 0, origin: { x: 0, y: 0, z: 0 } },
+        { buffer: dst, bytesPerRow: BYTES_PER_ROW },
+        { width: 3, height: 3, depthOrArrayLayers: 1 }
+      );
+      t.device.queue.submit([encoder.finish()]);
+
+      // Red should go 0, 127, 255 from left to right. Blue should to the same
+      // from top to bottom.
+      const expectedRows = [
+        [0x00, 0x00, 0x00, 0xff, 0x7f, 0x00, 0x00, 0xff, 0xff, 0x00, 0x00, 0xff],
+        [0x00, 0x00, 0x7f, 0xff, 0x7f, 0x00, 0x7f, 0xff, 0xff, 0x00, 0x7f, 0xff],
+        [0x00, 0x00, 0xff, 0xff, 0x7f, 0x00, 0xff, 0xff, 0xff, 0x00, 0xff, 0xff],
+      ];
+      for (const row of [0, 1, 2]) {
+        t.expectGPUBufferValuesPassCheck(
+          dst,
+          data => checkElementsEqual(data, new Uint8Array(expectedRows[row])),
+          { srcByteOffset: row * 256, type: Uint8Array, typedLength: 12 }
+        );
+      }
+    };
+
+    // If any iteration takes longer than this, we stop incrementing along that
+    // branch and move on to the next instance count. Note that the max
+    // supported vertex count for any iteration is 2**24 due to our choice of
+    // index buffer size.
+    const maxDurationMs = 100;
+    const counts: { numInstances: number; vertexCounts: number[] }[] = [
+      {
+        numInstances: 4,
+        vertexCounts: [2 ** 10, 2 ** 16, 2 ** 18, 2 ** 20, 2 ** 22, 2 ** 24],
+      },
+      {
+        numInstances: 2 ** 8,
+        vertexCounts: [2 ** 10, 2 ** 16, 2 ** 18, 2 ** 20, 2 ** 22],
+      },
+      {
+        numInstances: 2 ** 10,
+        vertexCounts: [2 ** 8, 2 ** 10, 2 ** 12, 2 ** 16, 2 ** 18, 2 ** 20],
+      },
+      {
+        numInstances: 2 ** 16,
+        vertexCounts: [2 ** 4, 2 ** 8, 2 ** 10, 2 ** 12, 2 ** 14],
+      },
+      {
+        numInstances: 2 ** 20,
+        vertexCounts: [2 ** 4, 2 ** 8, 2 ** 10],
+      },
+    ];
```

Comment:
I think this would work, and save a not-very-useful type annotation:

```suggestion
    const counts = [
      {
        numInstances: 4,
        vertexCounts: [2 ** 10, 2 ** 16, 2 ** 18, 2 ** 20, 2 ** 22, 2 ** 24],
      },
      {
        numInstances: 2 ** 8,
        vertexCounts: [2 ** 10, 2 ** 16, 2 ** 18, 2 ** 20, 2 ** 22],
      },
      {
        numInstances: 2 ** 10,
        vertexCounts: [2 ** 8, 2 ** 10, 2 ** 12, 2 ** 16, 2 ** 18, 2 ** 20],
      },
      {
        numInstances: 2 ** 16,
        vertexCounts: [2 ** 4, 2 ** 8, 2 ** 10, 2 ** 12, 2 ** 14],
      },
      {
        numInstances: 2 ** 20,
        vertexCounts: [2 ** 4, 2 ** 8, 2 ** 10],
      },
    ] as const;
```

(the `as const` might not even be necessary, not sure)

Reply:
Done